src/scanspec/specs.py

```diff
@@ -84,34 +89,42 @@ def shape(self) -> tuple[int, ...]:
         """Return the final, simplified shape of the scan."""
         return tuple(len(dim) for dim in self.calculate())
 
-    def __rmul__(self, other) -> Product[Axis]:
+    def __rmul__(self, other: int) -> Product[Axis]:
         return if_instance_do(other, int, lambda o: Product(Repeat(o), self))
 
-    def __mul__(self, other) -> Product[Axis]:
+    @overload
+    def __mul__(self, other: Spec[Axis]) -> Product[Axis]: ...
+
+    @overload
+    def __mul__(self, other: Spec[OtherAxis]) -> Product[Axis | OtherAxis]: ...
```

Comment:
It's variance like this that makes me think `duration` should be special, otherwise we end up with lots of `Spec[Motor | str]` and `isinstance` checks.

Alternatively we might decide that we always need `Spec[str]` with a lookup, as the case for `Spec[Motor]` in ophyd_async is looking tenuous. One for the scanspec v1 ticket (I realised we never made scanspec 1.0 so might as well use that number).

Reply:
A few thoughts here:

- `Spec[Motor]` would be very nice for blueapi, if we can get it to work with pydantic, even `Spec[Motor, str]`. 
- If we make time special there might be a way to incorporate it into a `Spec[Motor]` without making it a `Spec[Motor, str]` (or really a `Spec[Motor, Time]`, which would be preferable)
- _However_ we should ask ourselves if the variance is actually a bad thing, the extra verbosity here is just capturing a special case at "compile" time rather than runtime, the special case always existed and was always special, strict mode simply forces us to acknowledge it. In other words it is possible that a bunch of `isinstance` checks really might be the right thing to do here, or it's possible that we can box the specialness into one place by treating time as special somehow. I am not sure, but I don't think we should be prescriptive.